src/Uno.Extensions.Reactive/Core/State.Extensions.cs

```diff
@@ -199,9 +247,35 @@ public static IDisposable ForEachAsync<T>(this IState<T> state, AsyncAction<T?>
 	/// <param name="caller"> For debug purposes, the name of this subscription. DO NOT provide anything here, let the compiler fulfill this.</param>
 	/// <param name="line">For debug purposes, the name of this subscription. DO NOT provide anything here, let the compiler fulfill this.</param>
 	/// <returns>A <see cref="IDisposable"/> that can be used to remove the callback registration.</returns>
-	public static IDisposable ForEachAsync<T>(this IState<T?> state, AsyncAction<T?> action, [CallerMemberName] string? caller = null, [CallerLineNumber] int line = -1)
+	public static IState<T?> ForEach<T>(this IState<T?> state, AsyncAction<T?> action, [CallerMemberName] string? caller = null, [CallerLineNumber] int line = -1)
 		where T : struct
-		=> new StateForEach<T?>(state, action.SomeOrNone(), $"ForEachAsync defined in {caller} at line {line}.");
+	{
+		_ = AttachedProperty.GetOrCreate(owner: state,
+										 key: action,
+										 factory: (s, ks) => new StateForEach<T?>(state, action.SomeOrNone(), $"ForEach defined in {caller} at line {line}."));
+
+		return state;
+	}
+
+	/// <summary>
+	/// Execute an async callback each time the state is being updated.
+	/// </summary>
+	/// <typeparam name="T">The type of the state</typeparam>
+	/// <param name="state">The state to listen.</param>
+	/// <param name="action">The callback to invoke on each update of the state.</param>
+	/// <param name="caller"> For debug purposes, the name of this subscription. DO NOT provide anything here, let the compiler fulfill this.</param>
+	/// <param name="line">For debug purposes, the name of this subscription. DO NOT provide anything here, let the compiler fulfill this.</param>	
+	/// <param name="disposable"> A <see cref="IDisposable"/> that can be used to remove the callback registration.</param>
+	/// <returns>An <see cref="IState"/> that can be used to chain other operations.</returns>
+	public static IState<T?> ForEach<T>(this IState<T?> state, AsyncAction<T?> action, out IDisposable disposable, [CallerMemberName] string? caller = null, [CallerLineNumber] int line = -1)
+		where T : struct
+	{
+		disposable = AttachedProperty.GetOrCreate(owner: state,
+												  key: action,
+												  factory: (s, ks) => new StateForEach<T?>(state, action.SomeOrNone(), $"ForEach defined in {caller} at line {line}."));
+
+		return state;
+	}
 
 
 	/// <summary>
```

Comment:
We should also rename the `ForEachDataAsync` to `ForEachData`, and make it fluent